### PR TITLE
Add non-PWM examples

### PR DESCRIPTION
**Pull request by Tyler-Duckworth**
Somewhat fixes Issue #9 
I adapted all of the REV examples with the exception of one, the PWM Arcade Drive since there is not a PWM SparkMax object. 

I would note, I did get a lot of errors, but I think that is because the API is incomplete.

### CODE REVIEW

File: examples/can-arcade-drive/robot.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
```
Comment (auscompgeek) [inline examples/can-arcade-drive/robot.py]: If you're going to put a #! line here, you should make the file executable.

File: examples/get-set-params/robot.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        # you know if the parameter was successfully set:
+        #   CANError.kOk
+        #   CANError.kError
+        #   CANError.kTimeout
+        if self.motor.setIdleMode(rev.IdleMode.kCoast) != rev.CANError.kOK:
+            wpilib.SmartDashboard.putString("Idle Mode", "Error")
+
+        # Similarly, parameters will have a get() method which allows you to
+        # retrieve their values from the controller
+        if self.motor.getIdleMode() == rev.IdleMode.kCoast:
```
Comment (auscompgeek) [inline examples/get-set-params/robot.py]:
```suggestion
        if self.motor.getIdleMode() is rev.IdleMode.kCoast:
```

File: examples/position-pid-control/robot.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            self.coeff["d"] = d
+        if iz is not self.coeff["iz"]:
+            self.pidController.setIZone(iz)
+            self.coeff["iz"] = iz
+        if ff is not self.coeff["ff"]:
+            self.pidController.setFF(ff)
+            self.coeff["ff"] = ff
+        if (min_out is not self.kMinOutput) or (max_out is not self.kMaxOutput):
+            self.pidController.setOutputRange(min_out, max_out)
+            self.kMinOutput = min_out, self.kMaxOutput = max_out
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: This is not valid Python:
```pytb
  File "robot.py", line 84, in teleopPeriodic
    self.kMinOutput = min_out, self.kMaxOutput = max_out
TypeError: cannot unpack non-iterable float object
```
Please actually test your code.

File: examples/get-set-params/robot.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        # Parameters can be set by calling the appropriate set() method on the
+        # CANSparkMax object whose properties you want to change
+        #
+        # Set methods will return one of three CANError values which will let
+        # you know if the parameter was successfully set:
+        #   CANError.kOk
+        #   CANError.kError
+        #   CANError.kTimeout
+        if self.motor.setIdleMode(rev.IdleMode.kCoast) != rev.CANError.kOK:
```
Comment (auscompgeek) [inline examples/get-set-params/robot.py]:
```suggestion
        if self.motor.setIdleMode(rev.IdleMode.kCoast) is not rev.CANError.kOK:
```

File: examples/get-set-params/robot.py (hunk 2 above)
Comment (auscompgeek) [inline examples/get-set-params/robot.py]:
```suggestion
        if self.motor.getIdleMode() is rev.IdleMode.kCoast:
```

File: examples/get-set-params/robot.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        # Similarly, parameters will have a get() method which allows you to
+        # retrieve their values from the controller
+        if self.motor.getIdleMode() == rev.IdleMode.kCoast:
+            wpilib.SmartDashboard.putString("Idle Mode", "Coast")
+        else:
+            wpilib.SmartDashboard.putString("Idle Mode", "Brake")
+
+        # Set ramp rate to 0
+        if self.motor.setOpenLoopRampRate(0) != rev.CANError.kOK:
```
Comment (auscompgeek) [inline examples/get-set-params/robot.py]: Ditto.

File: examples/position-pid-control/robot.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        i = wpilib.SmartDashboard.getNumber("I Gain", 0)
+        d = wpilib.SmartDashboard.getNumber("D Gain", 0)
+        iz = wpilib.SmartDashboard.getNumber("I Zone", 0)
+        ff = wpilib.SmartDashboard.getNumber("Feed Forward", 0)
+        min_out = wpilib.SmartDashboard.getNumber("Min Output", 0)
+        max_out = wpilib.SmartDashboard.getNumber("Max Output", 0)
+        rotations = wpilib.SmartDashboard.getNumber("Set Rotations", 0)
+
+        # Update PIDController datapoints with the latest from SmartDashboard
+        if p is not self.coeff["p"]:
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: These all need to be `!=`.

File: examples/position-pid-control/robot.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        # You must call getPIDController() on an existing CANSparkMax or
+        # SparkMax object to fully use PID functionality
+        self.pidController = self.motor.getPIDController()
+
+        # Instantiate built-in encoder to display position
+        self.encoder = self.motor.getEncoder()
+
+        # PID Coefficents and Controller Output Range
+        self.coeff = {"p": 0.1, "i": 1e-4, "d": 0, "iz": 0, "ff": 0}
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: The integral zone is not a coefficient...

REV didn't write their examples like this either. Is there a particular reason why you didn't just copy and paste the existing Python code for this?

File: examples/position-pid-control/robot.py (hunk 3 above)
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: This is not valid Python:
```pytb
  File "robot.py", line 84, in teleopPeriodic
    self.kMinOutput = min_out, self.kMaxOutput = max_out
TypeError: cannot unpack non-iterable float object
```
Please actually test your code.

**Review by auscompgeek — CHANGES_REQUESTED** [review]
Why did you add this `.DS_Store` file?

File: examples/position-pid-control/robot.py (hunk 7 above)
Comment (Tyler-Duckworth) [reply examples/position-pid-control/robot.py]: Apologies for the errors. I didn't think twice after finishing them. Thanks for the feedback.

**Comment by virtuald** [comment]
Did you actually test the examples you created? We can't be giving high school students examples that don't work.

**Comment by Tyler-Duckworth** [comment]
@virtuald Yes, I did test them. The largest issue I had was distinguishing what errors were and were not a product of the API. I found a couple of functions, enums, etc. that I can detail later that weren't implemented in simulations, but worked fine on the RoboRIO. If there is anything you find that show them to be incorrect, please let me know so I can fix it.

File: examples/velocity-pid-control/robot.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        # rev.ControlType.kVelocity
+        # rev.ControlType.kVoltage
+        #
+        # For more information on what these types are, refer to the Spark Max
+        # documentation.
+        self.pidController.setReference(self.setpoint, rev.ControlType.kVelocity)
+
+        # Push Setpoint and the motor's current position to SmartDashboard.
+        wpilib.SmartDashboard.putNumber("Setpoint", self.setpoint)
+        wpilib.SmartDashboard.putNumber("Process Variable", self.encoder.getPosition())
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: This doesn't seem correct.

File: examples/velocity-pid-control/robot.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            self.kIz = iz
+        if ff != self.kFF:
+            self.pidController.setFF(ff)
+            self.kFF = ff
+        if (min_out != self.kMinOutput) or (max_out != self.kMaxOutput):
+            self.pidController.setOutputRange(min_out, max_out)
+            self.kMinOutput = min_out
+            self.kMaxOutput = max_out
+
+        self.setpoint = self.maxRPM * self.joystick.getY()
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: Why `self.setpoint`?

Comment (Tyler-Duckworth) [reply examples/velocity-pid-control/robot.py]: Just a preference, but I'll change it.

File: examples/velocity-pid-control/robot.py (hunk 8 above)
Comment (Tyler-Duckworth) [reply examples/velocity-pid-control/robot.py]: Got it. Thanks for pointing that out.

File: examples/can-arcade-drive/robot.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    #
+    # The example below initializes four brushless motors with CAN IDs
+    # 1, 2, 3 and 4. Change these parameters to match your setup
+
+    def robotInit(self):
+        # Create motors
+        self.leftLeadMotor = rev.CANSparkMax(1, rev.MotorType.kBrushless)
+        self.rightLeadMotor = rev.CANSparkMax(0, rev.MotorType.kBrushless)
+        self.leftFollowMotor = rev.CANSparkMax(2, rev.MotorType.kBrushless)
+        self.rightFollowMotor = rev.CANSparkMax(3, rev.MotorType.kBrushless)
```
Comment (auscompgeek) [inline examples/can-arcade-drive/robot.py]: The comment above lists CAN IDs 1, 2, 3, 4. Is there a reason why these are different?

Also, the comment above should be next to this code like it is in REV's examples.

File: examples/position-pid-control/robot.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+import rev
+import wpilib
+from wpilib.drive import DifferentialDrive
+
+
+class Robot(wpilib.TimedRobot):
+    def robotInit(self):
+        # Create motors
+        self.motor = rev.CANSparkMax(0, rev.MotorType.kBrushless)
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: Any particular reason why this is ID 0 (and not 1 like it is in REV's example)?

File: examples/read-encoder-values/robot.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    # Position is displayed in revolutions (of the motor's axle) and velocity
+    # is displayed in revolutions per minute (RPM)
+    #
+    # Optionally, if you call the setPositionConversionFactor() method on the
+    # encoder and give it a measurement of how far one revolution is, the
+    # getVelocity() and getPosition() methods return a scaled output in the
+    # units of your choice.
+    def robotInit(self):
+        # Instantiate SPARK MAX object
+        self.motor = rev.CANSparkMax(0, rev.MotorType.kBrushless)
```
Comment (auscompgeek) [inline examples/read-encoder-values/robot.py]: Ditto.

File: examples/velocity-pid-control/robot.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+import rev
+import wpilib
+from wpilib.drive import DifferentialDrive
+
+
+class Robot(wpilib.TimedRobot):
+    def robotInit(self):
+        # Create motors
+        self.motor = rev.CANSparkMax(0, rev.MotorType.kBrushless)
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: Ditto.

**Review by auscompgeek — COMMENTED** [review]
REV appear to have some Shuffleboard layouts to go with their examples too; looks like they have graphs or something. Mind adding them here too and checking that they work?

Also, is there a particular reason why you didn't port over the tank drive example too?

**Comment by auscompgeek** [comment]
Every enum available in the Java API should be available both in simulation and on the roboRIO. From my knowledge, the only methods that have yet to be implemented in simulation are follow and the low-level config API. Let us know if this isn't the case.

File: examples/can-arcade-drive/robot.py (hunk 10 above)
Comment (Tyler-Duckworth) [reply examples/can-arcade-drive/robot.py]: Sorry about that. Usually, when I configure my robots, their CAN IDs are zero-based.

**Comment by Tyler-Duckworth** [comment]
The main reason I didn't port over the Tank Drive example is that I did not know it was there. I was using the C++ examples for reference, but thanks for pointing that out. (The same goes for the Shuffleboard JSON files) As far as the CAN IDs go, usually when I write robot code I configure my CAN IDs to be zero-based like arrays. 

I was having some trouble with two main things: `rev.IdleMode` and `rev.CANSparkMax.restoreFactoryDefaults()` (Along with `rev.CANSparkMax.follow()`, but you already mentioned that)

**Comment by auscompgeek** [comment]
> As far as the CAN IDs go, usually when I write robot code I configure my CAN IDs to be zero-based like arrays.

0 isn't a valid CAN ID for the SPARK MAX API...

**Comment by auscompgeek** [comment]
> I was having some trouble with two main things: `rev.IdleMode`

That one definitely exists in simulation. Does it not for you?

> and `rev.CANSparkMax.restoreFactoryDefaults()`

That one should work fine. There _was_ a bug that would glitch out the simulation UI though that's fixed in master.

**Comment by Tyler-Duckworth** [comment]
I didn't know that CAN doesn't start at 0 for the Spark Max. Thanks!
Also, I figured out my issue with the simulation. I was testing on a laptop that didn't have admin privileges, so I think I had some issues with installing PIP packages. When I tried it on a computer with admin, it all worked.

**Comment by auscompgeek** [comment]
> I was testing on a laptop that didn't have admin privileges, so I think I had some issues with installing PIP packages. When I tried it on a computer with admin, it all worked.

pip works fine without admin privileges; I _never_ run pip as root/admin.

File: examples/limit-switch/robot.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+        # The get() method can be used on a CANDigitalInput object to read the
+        # state of the switch.
+        #
+        # In this example, the polarity of the switches are set to normally
+        # closed. In this case, get() will return true if the switch is
+        # pressed. It will also return true if you do not have a switch
+        # connected. get() will return false when the switch is released.
+        wpilib.SmartDashboard.putBoolean(
+            "Forward Limit Enabled", self.forwardLimit.get()
```
Comment (auscompgeek) [inline examples/limit-switch/robot.py]: This seems very wrong.

File: examples/position-pid-control/robot.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        # rev.ControlType.kVelocity
+        # rev.ControlType.kVoltage
+        #
+        # For more information on what these types are, refer to the Spark Max
+        # documentation.
+        self.pidController.setReference(rotations, rev.ControlType.kPosition)
+
+        # Push Setpoint and the motor's current position to SmartDashboard.
+        wpilib.SmartDashboard.putNumber("Setpoint", rotations)
+        wpilib.SmartDashboard.putNumber("Process Variable", self.encoder.getPosition())
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: These don't look like they match what's defined in the Shuffleboard layout below; did you actually test this?

File: examples/velocity-pid-control/robot.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        # rev.ControlType.kVelocity
+        # rev.ControlType.kVoltage
+        #
+        # For more information on what these types are, refer to the Spark Max
+        # documentation.
+        self.pidController.setReference(setpoint, rev.ControlType.kVelocity)
+
+        # Push Setpoint and the motor's current velocity to SmartDashboard.
+        wpilib.SmartDashboard.putNumber("Setpoint", setpoint)
+        wpilib.SmartDashboard.putNumber("Process Variable", self.encoder.getVelocity())
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: Ditto.

**Review by auscompgeek — DISMISSED** [review]
Please actually test your code. As Dustin said above, we _cannot_ be giving high school students broken code.

File: examples/velocity-pid-control/robot.py [hunk 17]
```diff
@@ -40,7 +40,7 @@ def robotInit(self):
         # configuration parameters in the SPARK MAX to their factory default
         # state. If no argument is passed, these parameters will not persist
         # between power cycles
-        self.motor.restoreFactoryDefaults()
+        # self.motor.restoreFactoryDefaults()
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: ...

**Comment by Tyler-Duckworth** [comment]
Got it. I do need to be more cognizant of undoing what I change to properly test these things on my computer. 

My apologies for the abundance of errors in this PR. I do need to be much more careful when writing code, especially when it is going to be used by other people.

File: examples/position-pid-control/robot.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2017-2018 FIRST. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+# ----------------------------------------------------------------------------
+
+import rev
+import wpilib
+from wpilib.drive import DifferentialDrive
```
Comment (auscompgeek) [inline examples/position-pid-control/robot.py]: examples/position-pid-control/robot.py:10:1: F401 'wpilib.drive.DifferentialDrive' imported but unused

File: examples/velocity-pid-control/robot.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2017-2018 FIRST. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+# ----------------------------------------------------------------------------
+
+import rev
+import wpilib
+from wpilib.drive import DifferentialDrive
```
Comment (auscompgeek) [inline examples/velocity-pid-control/robot.py]: examples/velocity-pid-control/robot.py:10:1: F401 'wpilib.drive.DifferentialDrive' imported but unused

**Review by auscompgeek — COMMENTED** [review]
(Hmm, we should probably run pyflakes in CI...)

**Comment by Tyler-Duckworth** [comment]
Is there anything more you need me to do? I can squash the commits into one, which would make merging easier.

**Comment by virtuald** [comment]
@Tyler-Duckworth sorry this took so long to approve, I wasn't really doing robotics stuff in the late summer. However, 2020 is upon us, and I just pushed the first 2020 version to pypi/roborio tonight. It would be great if you could try some of these on 2020.

Unfortunately, REV doesn't really support simulation... but I think I would be fine with them as long as they start and don't crash.

**Comment by auscompgeek** [comment]
For the PID examples, I would also appreciate it if the PID constants were double-checked against the current upstream. I think some of them may have been updated since this was opened.

**Comment by Tyler-Duckworth** [comment]
@virtuald It's fine! I do want to apologize for the poor first state of my PR. Thanks so much for the feedback about this PR. 

I definitely will try out the 2020 version. My team should be handing off a drivetrain with NEOs to me in the next few days, so I will try it then.

**Comment by auscompgeek** [comment]
Hey @Tyler-Duckworth, we'd love to finally get this merged in. What's the status on this?

**Comment by Tyler-Duckworth** [comment]
It's been a while since I've updated this with the latest from RobotPy, so there are probably some bugs I need to work out. I'll try to get to them this week!

**Pull request merged**